### PR TITLE
feat(PPDSC-2039): audio player event handler

### PR DESCRIPTION
**Pull request by Xin00163**
PPDSC-2039

Allow consumers to pass an extra function which would be called before our event handlers.

**I have done:**
 - [ ] Written unit tests against changes
 - [ ] Written accessibility tests against the component
 - [ ] Written functional tests against the component and/or NewsKit site
 - [ ] Updated relevant documentation
 - [ ] Updated the release notes for the [next release version](https://nidigitalsolutions.jira.com/wiki/spaces/NPP/pages/2349695021/Release%2BNotes%2BCode%2B-%2BWeb%2BDocumentation)
 - [ ] Discussed with QA and/or design about any visual changes

**I have tested manually:**
 - [ ] Renders correctly on IE11
 - [ ] The feature's functionality is working as expected on other browsers such as Safari, Firefox and IE11
 - [ ] There are no accessibility errors reported by WAVE
 - [ ] The screen reader reads the expected text for the tabbable elements.
 - [ ] There are no new errors in the browser console coming from this PR.
 - [ ] When visual test is not added, it renders correctly on different browsers and mobile viewports (Safari, Firefox, small mobile viewport, tablet)
 - [ ] The Playground feature is working as expected

**Before**

**After:**

**Who should review this PR**

**How to test**

### CODE REVIEW

File: src/audio-player-composable/audio-player-composable.tsx [hunk 1]
```diff
@@ -119,18 +120,18 @@ export const AudioPlayerComposable = ({
...
   return (
     <section aria-label={ariaLandmark || 'Audio Player'}>
       <AudioPlayerProvider value={value}>
-        <AudioElement />
+        <AudioElement
+          audioRef={audioRef}
+          audioEvents={audioEvents}
+          src={src}
+          autoPlay={autoPlay}
+          {...props}
```
Comment (Vanals) [inline src/audio-player-composable/audio-player-composable.tsx:133]: Wonder if is better to destructure the event props ,it is then  going to be more readable and we avoid passing unnecessary/wrong attributes to the audio element. 

Same as we are doing with the `src`. I know they are many but I think is worth

Comment (Xin00163) [reply src/audio-player-composable/audio-player-composable.tsx]: Make sense to me. Have implemented the changes. Pls let me know if they look right

**Review by Vanals — APPROVED** [review]

File: src/audio-player-composable/audio-player-composable.tsx [hunk 2]
```diff
@@ -119,18 +120,32 @@ export const AudioPlayerComposable = ({
...
-    audioRef,
-    audioEvents,
-    src,
-    autoPlay,
+  const eventHandler = (eventName: AudioEvents) => (
+    e: React.SyntheticEvent<HTMLAudioElement, Event>,
+  ) => {
+    const eventCallback = props[eventName];
+    // eslint-disable-next-line @typescript-eslint/no-unused-expressions
+    eventCallback ? eventCallback(e) : audioEvents![eventName](e);
```
Comment (mutebg) [inline src/audio-player-composable/audio-player-composable.tsx:130]: this is slightly incorrect, we need to combine our internal audio event with the user-provided one if exist.
right now if I add onDurationChange it breaks the player.

In order to combine events you can use the util -> composeEventHandlers;
```
const eventHandler = (eventName: AudioEvents) => {
    const propEvent = props[eventName];
    const internalEvent = audioEvents[eventName];
    return composeEventHandlers([propEvent, internalEvent]);
  };
  ```

**Review by mutebg — CHANGES_REQUESTED** [review]

File: src/audio-player-composable/audio-player-composable.tsx [hunk 3]
```diff
@@ -147,14 +152,18 @@ export const AudioPlayerComposable = ({
...
   const eventHandler = (eventName: AudioEvents) => (
     e: React.SyntheticEvent<HTMLAudioElement, Event>,
   ) => {
-    const eventCallback = props[eventName];
-    // eslint-disable-next-line @typescript-eslint/no-unused-expressions
-    eventCallback ? eventCallback(e) : audioEvents![eventName](e);
+    const propEvent = props[eventName];
+    const internalEvent = audioEvents[eventName];
+    return propEvent
+      ? composeEventHandlers([propEvent(e), internalEvent(e)])
```
Comment (mutebg) [inline src/audio-player-composable/audio-player-composable.tsx:165]:
```js
const eventHandler = (event) => {
    const propEvent = props[eventName];
    const internalEvent = audioEvents[eventName];
    return composeEventHandlers([propEvent, internalEvent])
```

**Review by mutebg — APPROVED** [review]

**Review by Vanals — DISMISSED** [review]

**Review by mutebg — APPROVED** [review]

**Review by Vanals — APPROVED** [review]

**Pull request merged**